File: 206. Reverse Linked List/ReverseList.java
/**
 * https://leetcode.com/problems/reverse-linked-list/
 */

public class ReverseList {
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next;
        }
    }

    public ListNode reverseList(ListNode head) {
        ListNode next;
        ListNode prev = head;
        ListNode curr = head.next;
        ListNode orgHead = head;

        while (curr != null)
        {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        orgHead = null;
        head = prev;

        return head;
    }
}

File: 206. Reverse Linked List/ReverseList.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    # O(n) time,
    # O(1) space,
    # Approach: linked list, two pointers
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        prev, curr = head, head.next
        prev.next = None
        while curr:
            nxt = curr.next
            curr.next = prev

            prev = curr
            curr = nxt

        return prev
